### DISCUSSION TITLE
How do I handle a BlockGap/Margin issue?

### DISCUSSION THREAD

**Discussion by CGTS** [WordPress/gutenberg#47935] [Q&A]
In my `theme.json` file when I set `"blockGap": true,` under "settings". I get the default spacing in my elements that contain `gap` in the css and I can set defaults for individual blocks as expected.

When I set  `"blockGap": false,` I lose the default spacing on my elements and I am still able to set defaults for individual blocks. This also works as expected.

When I don't include blockgap in the `theme.json` file or choose `"blockGap": null"` I lose the default spacing and the ability to set default for blocks individually. Not quite what I was expecting but I can understand how that might work.

What I can't figure out is why choosing either `"blockGap": true,` or `"blockGap": false,` will add

```
.wp-site-blocks > * + * {
    margin-block-start: 1.5rem;
}

.wp-site-blocks > * + * {
    -webkit-margin-before: inherit;
    margin-block-start: inherit;
}
```

This is causing spacing issues between each high-level section and creates a stair-step effect on a block where I have a custom grid. I've tried overwriting it in the CSS

```
.wp-site-blocks > * + * { margin-block-start: inherit; }
.editor-styles-wrapper .wp-site-blocks > * + * { margin-block-start: inherit !important; }
```

It works on the front-end but not in the editor. I also tried to dequeue the site-blocks styles but was unsuccessful. 

Is there any way to get rid of the margins added to the high-level blocks when I turn blockGap on. I've see some discussion in the issues area but not solutions for dealing with it.

**Comment by CGTS** [comment]
I did some more testing and [reading](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/). I ended up just setting `"blockGap": "0"` in `theme.json` and will set the gap spacing per block. Not ideal but will work. I'm still open to ways to remove the margin on each block through `theme.json` or removing the style if anyone has a suggestion.

**Comment by tresorama** [comment]
I've not tested `blockGap` intensively. I'm new to block themes.
But coming from a React background, where i usually use Tailwind i find this `blockGap` weird.
But maybe i've not used enough to understand its benefits.

<hr/>

In my local test with block theme, i used TT3 (twenty-twenty-three theme) as starting point and play a bit with theme.json.

<hr/>

I noticed that this global `blockGap` defined in `theme.json` is used in `core/navigation` component/block to separate nav items.
It is also used to separate root `template` sections, meaning component/block in the `template` that has no parent.

First weirdness, 
> why these two separte thing need a shared value ???

Second weirdness,
Imagining that i want that root sections of my page has a gap between each other i would change this pseudo React Tree
```tsx
<SectionHero/>
<SectionTestimonials/>
<SectionCTA/>
```
to this
```tsx
<Stack itemsGap={...}>
  <SectionHero/>
  <SectionTestimonials/>
  <SectionCTA/>
</Stack>
```

Doing like that i will add this `<Stack>` only in page that requires it.
Doing like that i can have `<Stack itemsGap={10px}>` in one page and `<Stack itemsGap={60px}>`  in other page.
I found this predictable and easy.

I don't see the point of anticipate this at whole app level.
If i decide a `blockGap` at global level - higher level than single template - this means that every template/page will share the same gap between sections.
How could i disable it at page level ?
How could i disable it at single block/components level ?
What wrong with using padding and margin where there is necessary ?

<hr/>

In my test I ended up setting `blockGap` to `0px` and that i used layout components/blocks (Group, Row, Stack) and block props (Padding Margin) where needed.

So IMO, your...

> I did some more testing and [reading](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/). I ended up just setting "blockGap": "0" in theme.json and will set the gap spacing per block. Not ideal but will work. I'm still open to ways to remove the margin on each block through theme.json or removing the style if anyone has a suggestion.

... sound ideal for me.

Just my 2 cents.
